parser/ParseIndexProperties.hpp

```diff
@@ -0,0 +1,271 @@
+ */
+class IndexProperties {
+ public:
+  typedef std::unordered_map<std::string, ParseKeyValue::KeyValueType> valid_property_map_t;
+
+  bool hasValidIndexDescription() const {
+    return is_index_description_valid_;
+  }
+
+  const ParseKeyValue* getInvalidPropertyNode() const {
+    return invalid_property_node_;
+  }
+
+  std::string getReasonForInvalidIndexDescription() const {
+    return reason_for_invalid_description_;
+  }
+
+  const IndexSubBlockDescription* getIndexDescription() const {
+    return index_sub_block_description_.get();
+  }
+
+  void setIndexDescriptionAsInvalid(std::string reason) {
+    is_index_description_valid_ = false;
+    index_sub_block_description_.reset();
+    reason_for_invalid_description_ = reason;
+    invalid_property_node_ = nullptr;
+  }
+
+  void setIndexDescriptionAsInvalid(const ParseKeyValue *invalid_node, std::string reason) {
+    is_index_description_valid_ = false;
+    index_sub_block_description_.reset();
+    reason_for_invalid_description_ = reason;
+    invalid_property_node_ = invalid_node;
+  }
+
+  virtual bool addCustomProperties(const PtrList<ParseKeyValue> *key_value_list) = 0;
+
+  bool hasDuplicateCustomProperty(const PtrList<ParseKeyValue> *key_value_list) {
+    // define an existence map, which checks for duplicate properties in the key_value_list
+    std::unordered_map<std::string, bool> is_duplicate_property;
+    for (const ParseKeyValue &key_value : *key_value_list) {
+      std::string key = ToLower(key_value.key()->value());
+      if (is_duplicate_property.find(key) != is_duplicate_property.end()) {
+        setIndexDescriptionAsInvalid(&key_value, key + " is specified more than once");
+        return true;  // duplicate found, return true
+      } else {
+        is_duplicate_property[key] = true;
+      }
+    }
+    return false;  // no duplicates, return false
+  }
+
+  bool hasInvalidCustomProperty(const PtrList<ParseKeyValue> *key_value_list) {
+    for (const ParseKeyValue &key_value : *key_value_list) {
+      std::string key = ToLower(key_value.key()->value());
+      if (valid_property_map_.find(key) == valid_property_map_.end()) {
+        setIndexDescriptionAsInvalid(&key_value, key + " is an invalid property for this index");
+        return true;  // this is not one of the valid keys, return true
+      } else {
+        // validate the type
+        if (key_value.getKeyValueType() != valid_property_map_[key]) {
+          setIndexDescriptionAsInvalid(&key_value, key + " property has an invalid type for this index");
+          return true;  // this key has an unexpected type, return true
+        }
+      }
+    }
+    return false;  // all valid keys, return false
+  }
+
+ protected:
+  valid_property_map_t valid_property_map_;  // a map of index property constants and their associated type
+  bool is_index_description_valid_;
+  const ParseKeyValue* invalid_property_node_;  // referred object owned by ParseIndexProperties class
```

Comment:
Pointer should move over to the var name: `const ParseKeyValue *invalid_property_node_;`